### PR TITLE
Add standalone `AnalyzerRule` example that implements row level access control

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?

Closes https://github.com/apache/datafusion/issues/10855

## Rationale for this change

It is common to use DataFusion to analyze / modify queries, so it would be good to have some more full featured examples of doing so.

Additionally,  the current `rewrite_expr.rs` example has three distinct examples:
1. Using DataFusion LogicalPlanning without execution
2. Implementing an `AnalyzerRule`
3. Implementing an `OptimizerRule`

Each I think should be its own example to make them easier to find and more full featured

## What changes are included in this PR?

1. Add example `analyzer_rule.rs` that implements "row level access control". I was inspired by @goldmedal 's description of Wren's usecase in https://github.com/apache/datafusion/issues/10855#issuecomment-2171124147 to pick the row level access control example

## Are these changes tested?
By CI

### CODE REVIEW

File: datafusion-examples/examples/analyzer_rule.rs [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+use std::sync::{Arc, Mutex};
+
+/// This example demonstrates how to add your own [`AnalyzerRule`] to
+/// DataFusion.
+///
+/// [`AnalyzerRule`]s transform [`LogicalPlan`]s prior to  the DataFusion
+/// optimization process, and can be used to change the plan's semantics (e.g.
+/// output types).
+///
+/// This example shows an `AnalyzerRule` which implements a  simplistic of row
```
Comment (goldmedal) [inline datafusion-examples/examples/analyzer_rule.rs:34]:
```suggestion
/// This example shows an `AnalyzerRule` which implements a simplistic of row
```
Redundant space.

**Review by goldmedal — APPROVED** [review]
Thanks @alamb -- I think this example clearly shows how to use the `AnalyzerRule` and `TreeNode` APIs. LGTM.

File: datafusion-examples/examples/analyzer_rule.rs [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+    // +----------+-----+----------+
+    // | name     | age | position |
+    // +----------+-----+----------+
+    // | Andrew   | 22  | Manager  |
+    // | Chunchun | 44  | Manager  |
+    // +----------+-----+----------+
+    rule.set_show_position("Manager");
+    ctx.sql("SELECT * FROM employee").await?.show().await?;
+
+    // The filters introduced by our AanalyzerRule are treated the same as any
```
Comment (jonahgao) [inline datafusion-examples/examples/analyzer_rule.rs:85]:
```suggestion
    // The filters introduced by our AnalyzerRule are treated the same as any
```

File: datafusion-examples/examples/analyzer_rule.rs [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+    // when we print out the plan
+    //
+    // Filter: employee.age > Int32(30) AND employee.position = Utf8("Manager")
+    //   TableScan: employee projection=[name, age, position]
+    println!("Logical Plan:\n\n{}\n", plan.display_indent());
+
+    Ok(())
+}
+
+/// Example AnalyzerRulw ule that implements a very basic "row level access
```
Comment (jonahgao) [inline datafusion-examples/examples/analyzer_rule.rs:105]:
```suggestion
/// Example AnalyzerRule that implements a very basic "row level access
```

**Review by jonahgao — APPROVED** [review]
Looks good to me!

**Comment by alamb** [comment]
Thank you for the review @jonahgao and @goldmedal

**Pull request merged**